bigchaindb/util.py

```diff
@@ -251,16 +373,87 @@ def verify_signature(signed_transaction):
         bool: True if the signature is correct, False otherwise.
     """
 
-    data = signed_transaction.copy()
+    for fulfillment in signed_transaction['transaction']['fulfillments']:
+        fulfillment_message = get_fulfillment_message(signed_transaction, fulfillment)
+        # verify the fulfillment (for now lets assume there is only one owner)
+        try:
+            parsed_fulfillment = cc.Fulfillment.from_uri(fulfillment['fulfillment'])
+        except Exception:
+            return False
+        is_valid = parsed_fulfillment.validate(serialize(fulfillment_message))
+
+        # if not a `CREATE` transaction
+        if fulfillment['input']:
+            is_valid &= parsed_fulfillment.condition.serialize_uri() == \
+                fulfillment_message['condition']['condition']['uri']
+        if not is_valid:
+            return False
+
+    return True
+
+
+def get_fulfillment_message(transaction, fulfillment, serialized=False):
+    """Get the fulfillment message for signing a specific fulfillment in a transaction
+
+    Args:
+        transaction (dict): a transaction
+        fulfillment (dict): a specific fulfillment (for a condition index) within the transaction
+        serialized (bool): False returns a dict, True returns a serialized string
+
+    Returns:
+        str|dict: fulfillment message
+    """
+    b = bigchaindb.Bigchain()
+
+    common_data = {
+        'operation': transaction['transaction']['operation'],
+        'timestamp': transaction['transaction']['timestamp'],
+        'data': transaction['transaction']['data'],
+        'version': transaction['version'],
+        'id': transaction['id']
+    }
+
+    fulfillment_message = common_data.copy()
+    fulfillment_message.update({
+        'input': fulfillment['input'],
+        'condition': None,
+    })
+
+    # if `TRANSFER` transaction
+    if fulfillment['input']:
+        # get previous condition
+        previous_tx = b.get_transaction(fulfillment['input']['txid'])
+        conditions = sorted(previous_tx['transaction']['conditions'], key=lambda d: d['cid'])
```

Comment:
They are already in an list but there are no guarantees on the order of the list.
I need to sort it so that I can access the correct condtion (or output) in a multi output transaction.
`conditions[fulfillment['input']['cid']]`